relay/relay.go

```diff
@@ -273,11 +275,36 @@ func (r *Relay) process(parentCtx context.Context) error {
 		}
 	}()
 
+	readerRetry, err := retry.NewReaderRetry(r.cfg.ReaderRetry)
+	if err != nil {
+		return err
+	}
+
 	transformer2 := transformer.NewTransformer(parser2)
 
 	go r.doIntervalOps(parentCtx)
 
-	return r.handleEvents(parentCtx, reader2, transformer2, writer2)
+	// handles binlog events with retry mechanism.
+	// it only do the retry for some binlog reader error now.
+	for {
+		err := r.handleEvents(parentCtx, reader2, transformer2, writer2)
+		if err == nil {
+			return nil
+		} else if !readerRetry.Check(parentCtx, err) {
+			return err
+		}
+
+		r.tctx.L().Info("receive retryable error for binlog reader", log.ShortError(err))
+		err = reader2.Close() // close the previous reader
+		if err != nil {
+			r.tctx.L().Error("fail to close binlog event reader", zap.Error(err))
+		}
+		reader2, err = r.setUpReader() // setup a new one
+		if err != nil {
+			return err
```

Comment:
does need it `errors.Trace`?

Reply:
the returned error is a `terror.Error` with `code`, `class`, `scope`, etc.
so, is it still need to use `errors.Trace`?

Reply:
it is not allowed to wrap `*terror.Error` with `errors.Trace`, which will lose error trace stack